packages/web-api/src/auth/routes.ts

```diff
@@ -32,20 +37,41 @@ import {
   getRefreshTokenObject,
   isRefreshTokenValid as validateRefreshToken
 } from './utils';
-import z from 'zod';
-import { PreApprovedUserService } from '../services/preApproval';
 
 require('express-async-errors');
 export const router: Router = express.Router();
 
 // All users are granted this role by default
 export const BASE_ROLES: UserRole[] = [UserRole.DEFAULT];
 
+// Rate limiting for password routes
+const authRateLimit = (per15Minutes: number) => {
+  // If in test mode, return a pass-through middleware
+  if (process.env.TEST_MODE === 'true') {
+    return (req: any, res: any, next: any) => next();
+  }
+
+  return rateLimit({
+    // 15 true
+    windowMs: 15 * 60 * 1000,
+    // configurable attempts within 15 minutes
+    max: per15Minutes,
+    keyGenerator: async req => {
+      // Rate limit by user ID then IP then unknown as backup
+      return req.user?.id.toString() ?? req.ip ?? 'unknown';
```

Comment:
I'm thinking throw error if `req.user.id` and `req.ip` both missing, something is very wrong if `req` has no properties.